fix: throwing an error when non-empty options were passed empty

KostiantynPopovych opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes #22732 

### Additional details
I hope this PR will greatly improve DX, as it is much easier to maintain applications where shared strings are taken from a single point and not recreated each time.

I'm unsure how properly update snapshots because some are related to the operation system (eg. prev: `Platform: test platform-x64 (Foo-OsVersion)`, current: `Platform: test platform-arm64 (Foo-OsVersion)`, but there is at least one snapshot change in the `cli/__snapshoots__/error_spec.js` file.

I will happily address any comments, suggestions, or criticisms and continue working on these improvements!


### Steps to test
I guess it is worth retesting all CLI commands and a couple of their combinations 🫣

### How has the user experience changed?
This PR will improve UX allowing users to understand what the problem is truly.

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=26440) <br/>All committers have signed the CLA.

emilyrohrbough commented on cli/lib/constants/options.js:
```diff
@@ -0,0 +1,120 @@
+  [camelNames.spec]: 'runs specific spec file(s). defaults to "all"',
+  [camelNames.tag]: 'named tag(s) for recorded runs in Cypress Cloud',
+  [camelNames.version]: 'prints Cypress version',
+}
```
@KostiantynPopovych Thank you for taking on this change! At a high-level glance, it appears this PR is tackling two things:
1. the bug where we report the wrong flag when a missing flag value was provided
2. refactoring to allow these up these flags to live in one place

This is might make this review a little slower, since a lot changes. My first pass looking at this is restructuring this data would make it a bit clearer to understand the rest of the changes. Something like the follow would be easier to reason with:

```js
const commandOptions = {
  'options_as_camelCase': {
     description:
     flag:
     shortHand:    
   }
}
```
so
```js
const commandOptions = {
  exit: {
      description: '...',
      flag: '--no-exit',
      // don't define shore flag since there isn't one.
  },
  reporterOptions: {
     description: '...',
     flag: '--reporter-options',
     shortFlag: '-o', 
   },
}
```

emilyrohrbough commented on cli/test/lib/exec/open_spec.js:
```diff
@@ -138,5 +138,9 @@ describe('exec open', function () {
     })
+
+    it('throws an error when passed an empty option that needs to have a value', () => {
+      expect(() => open.processRunOptions({ browser: '' })).to.throw()
```
What does this throw?

emilyrohrbough commented on cli/lib/exec/run.js:
```diff
@@ -96,58 +61,47 @@ const processRunOptions = (options = {}) => {
+    { option: camelNames.browser, condition: options.browser },
+    { option: camelNames.ciBuildId, condition: options.ciBuildId },
+    { option: camelNames.config, condition: options.config },
+    { option: camelNames.configFile, condition: options.configFile !== undefined, customProcessing: () => checkConfigFile(options) },
```
I think it'd be worth moving this check specifically out of this helper and reducing this to:
```js
const runOpts = [
  'browser',
  'ciBuildId',
]

runOpts.forEach((opt) => P
  if (options[opt]) {
     args.push(cypressOptions[op].flag)
  })
})

```

KostiantynPopovych commented:
Hey @emilyrohrbough, thank you for the review and your comments. I addressed them, so now we have a config with all needed options-related info (I hope). And also, related to your question about what throws that test - I changed the logic to store the result in the snapshot file, so now you can look at what it looks like - `cli/__snapshots__/open_spec.js`. Also, the `processOpenOptions` function was modified.

mschile commented on cli/lib/constants/options.js:
```diff
@@ -0,0 +1,84 @@
+const camelNames = {
```
```suggestion
const cliOptionKeys = {
```

mschile commented on cli/lib/exec/open.js:
```diff
@@ -27,42 +35,27 @@ const processOpenOptions = (options = {}) => {
-  if (options.config) {
-    args.push('--config', options.config)
-  }
+  const optionList = [
```
Instead of listing the options here, suggest adding another property to the option object that indicates if it's used in `open` mode, `run` mode, or both modes. Then you could just query for the options that apply to the particular mode.

mschile commented on cli/lib/exec/run.js:
```diff
@@ -96,59 +61,55 @@ const processRunOptions = (options = {}) => {
 
-  if (options.tag) {
-    args.push('--tag', options.tag)
+  if (options.autoCancelAfterFailures !== undefined) {
```
Was this intentionally changed from:
```js
if (options.autoCancelAfterFailures || options.autoCancelAfterFailures === 0 || options.autoCancelAfterFailures === false) {
```

mschile commented on cli/lib/exec/shared.js:
```diff
@@ -58,8 +59,52 @@ const checkConfigFile = (options) => {
+* @returns {void}
+* */
+const validateNonEmptyOptions = (options) => {
+  const nonEmptyOptions = [
```
Suggest moving this list into `options.js`. May make sense to move the entire function there as well.

chrisbreiding reviewed (changes requested):
@KostiantynPopovych Can you please revert the refactoring changes that aren't necessary for fixing the issue at hand? They cloud the actual changes for the fix. We might consider them as a separate PR, but honestly, I don't think they're a net positive.

KostiantynPopovych commented:
@chrisbreiding, sure, I'll do it during the next few days.

KostiantynPopovych committed:
Fix error message when the option that expects value was passed without it

KostiantynPopovych committed:
Merge branch 'develop' into issue-22732

KostiantynPopovych commented:
@chrisbreiding Reverted. Only a related fix is in place.

chrisbreiding committed:
Merge branch 'develop' into issue-22732

chrisbreiding commented:
@KostiantynPopovych Thanks for that! 

It looks like this is causing a test failure, so that will need to be addressed: https://app.circleci.com/pipelines/github/cypress-io/cypress/52457/workflows/d58d0bb7-feff-44ba-a20c-d50cf6326051/jobs/2165513

mschile commented:
@KostiantynPopovych, I tested the issue reported with #22732 and the error still occurs after these changes.

```
➜  cypress git:(issue-22732) ✗ yarn cypress:run --browser --spec cypress/e2e/1-getting-started/todo.cy.js
yarn run v1.22.19
$ yarn ensure-deps
$ ./scripts/ensure-dependencies.sh
success Folder in sync.
$ cypress run --dev --browser --spec cypress/e2e/1-getting-started/todo.cy.js
Cypress encountered an error while parsing the argument: --spec

You passed: true

The error was: spec must be a string or comma-separated list
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I believe the issue is earlier in the process when the arguments are actually being parsed.

mschile commented:
@KostiantynPopovych, unfortunately, based on this [comment](https://github.com/cypress-io/cypress/issues/22732#issuecomment-1553744698) I don't believe your approach is going to work. As such, I am going to close this PR. Sorry and thanks again for taking the time to open a PR.

KostiantynPopovych commented:
@mschile, thank you for clarifications and double-checking! Just one question/clarification, shouldn't we keep this fix (probably modify it) as a part of the API implementation? Is it okay to pass an empty string as an argument for the option which requires some value?

mschile commented:
@KostiantynPopovych, `commander` will throw an error if a required value is missing:

```
npx cypress run --browser
error: option '-b, --browser <browser-name-or-path>' argument missing
```
